core/src/main/java/com/novoda/noplayer/internal/exoplayer/drm/ExoPlayerDrmSessionCreator.java

```diff
@@ -0,0 +1,90 @@
+
+import androidx.annotation.Nullable;
+
+class ExoPlayerDrmSessionCreator implements DrmSessionCreator {
+
+    @SuppressWarnings("PMD.LooseCoupling")  // Unfortunately the DefaultDrmSessionManager takes a HashMap, not a Map
+    private static final HashMap<String, String> NO_OPTIONAL_PARAMETERS = null;
+
+    private final KeyRequestExecutor keyRequestExecutor;
+    private final ProvisionExecutor provisionExecutor;
+    @Nullable
+    private final KeySetId keySetId;
+    private final FrameworkMediaDrmCreator mediaDrmCreator;
+    private final Handler handler;
+
+    ExoPlayerDrmSessionCreator(KeyRequestExecutor keyRequestExecutor,
+                               ProvisionExecutor provisionExecutor,
+                               @Nullable KeySetId keySetId,
+                               FrameworkMediaDrmCreator mediaDrmCreator,
+                               Handler handler) {
+        this.keyRequestExecutor = keyRequestExecutor;
+        this.provisionExecutor = provisionExecutor;
+        this.keySetId = keySetId;
+        this.mediaDrmCreator = mediaDrmCreator;
+        this.handler = handler;
+    }
+
+    @Override
+    public DefaultDrmSessionManager<FrameworkMediaCrypto> create(DefaultDrmSessionEventListener eventListener) {
+        ProvisioningModularDrmCallback mediaDrmCallback = new ProvisioningModularDrmCallback(
+                keyRequestExecutor,
+                provisionExecutor
+        );
+
+        FrameworkMediaDrm frameworkMediaDrm = mediaDrmCreator.create(WIDEVINE_MODULAR_UUID);
+        DefaultDrmSessionManager<FrameworkMediaCrypto> drmSessionManager = new DefaultDrmSessionManager<>(
+                WIDEVINE_MODULAR_UUID,
+                frameworkMediaDrm,
+                mediaDrmCallback,
+                NO_OPTIONAL_PARAMETERS
+        );
+
+        if (keySetId != null) {
+            checkExpiryOfKeySetId(eventListener, drmSessionManager);
+        }
+
+        drmSessionManager.removeListener(eventListener);
+        drmSessionManager.addListener(handler, eventListener);
+
+        return drmSessionManager;
+    }
+
+    private void checkExpiryOfKeySetId(DefaultDrmSessionEventListener eventListener,
+                                       DefaultDrmSessionManager<FrameworkMediaCrypto> drmSessionManager) {
+        try {
+            OfflineLicenseHelper<FrameworkMediaCrypto> offlineLicenseHelper = OfflineLicenseHelper.newWidevineInstance(
+                    "",
+                    null
+            );
+
+            byte[] rawKeySetId = keySetId.asBytes();
+            Pair<Long, Long> licenseDurationRemainingSec = offlineLicenseHelper.getLicenseDurationRemainingSec(rawKeySetId);
+            Long first = licenseDurationRemainingSec.first;
+            if (first > TimeUnit.HOURS.toSeconds(1)) {
+                drmSessionManager.setMode(DefaultDrmSessionManager.MODE_QUERY, rawKeySetId);
+            }
+
+        } catch (UnsupportedDrmException | DrmSession.DrmSessionException e) {
```

Comment:
maybe for a next task, let's check if we actually handle properly those types of exceptions

Reply:
As per the docs on the `onDrmSessionManagerError`: 
```
  /**
   * Called when a drm error occurs.
   *
   * <p>This method being called does not indicate that playback has failed, or that it will fail.
   * The player may be able to recover from the error and continue. Hence applications should
   * <em>not</em> implement this method to display a user visible error or initiate an application
   * level retry ({@link Player.EventListener#onPlayerError} is the appropriate place to implement
   * such behavior). This method is called to provide the application with an opportunity to log the
   * error if it wishes to do so.
   *
   * @param error The corresponding exception.
   */
```